e2e/tests/wasm/upgrade_test.go

```diff
@@ -0,0 +1,157 @@
+	testvalues.VotingPeriod = time.Minute * 5
+
+	testifysuite.Run(t, new(IBCWasmUpgradeTestSuite))
+}
+
+type IBCWasmUpgradeTestSuite struct {
+	testsuite.E2ETestSuite
+}
+
+func (s *IBCWasmUpgradeTestSuite) TestIBCWasmChainUpgrade() {
+	t := s.T()
+
+	ctx := context.Background()
+	chain := s.SetupSingleChain(ctx)
+	checksum := ""
+
+	userWallet := s.CreateUserOnChainA(ctx, testvalues.StartingTokenAmount)
+	s.Require().NoError(testutil.WaitForBlocks(ctx, 1, chain), "failed to wait for blocks")
+
+	t.Run("create and exec store code proposal", func(t *testing.T) {
+		file, err := os.Open("contracts/ics10_grandpa_cw.wasm.gz")
+		s.Require().NoError(err)
+
+		checksum = s.ExecStoreCodeProposal(ctx, chain.(*cosmos.CosmosChain), userWallet, file)
+		s.Require().NotEmpty(checksum, "checksum must not be empty")
+	})
+
+	t.Run("upgrade chain", func(t *testing.T) {
+		testCfg := testsuite.LoadConfig()
+		s.UpgradeChain(ctx, chain.(*cosmos.CosmosChain), userWallet, testCfg.UpgradeConfig.PlanName, testCfg.ChainConfigs[0].Tag, testCfg.UpgradeConfig.Tag)
+	})
+
+	t.Run("query wasm checksums", func(t *testing.T) {
+		checksums, err := s.QueryWasmChecksums(ctx, chain)
+		s.Require().NoError(err)
+		s.Require().Contains(checksums, checksum)
+	})
+}
+
+// UpgradeChain upgrades a chain to a specific version using the planName provided.
+// The software upgrade proposal is broadcast by the provided wallet.
+func (s *IBCWasmUpgradeTestSuite) UpgradeChain(ctx context.Context, chain *cosmos.CosmosChain, wallet ibc.Wallet, planName, currentVersion, upgradeVersion string) {
+	plan := upgradetypes.Plan{
+		Name:   planName,
+		Height: int64(haltHeight),
+		Info:   fmt.Sprintf("upgrade version test from %s to %s", currentVersion, upgradeVersion),
+	}
+
+	upgradeProposal := upgradetypes.NewSoftwareUpgradeProposal(fmt.Sprintf("upgrade from %s to %s", currentVersion, upgradeVersion), "upgrade chain E2E test", plan)
+	s.ExecuteAndPassGovV1Beta1Proposal(ctx, chain, wallet, upgradeProposal)
+
+	height, err := chain.Height(ctx)
+	s.Require().NoError(err, "error fetching height before upgrade")
+
+	timeoutCtx, timeoutCtxCancel := context.WithTimeout(ctx, time.Minute*2)
+	defer timeoutCtxCancel()
+
+	err = testutil.WaitForBlocks(timeoutCtx, int(haltHeight-height)+1, chain)
+	s.Require().Error(err, "chain did not halt at halt height")
+
+	err = chain.StopAllNodes(ctx)
+	s.Require().NoError(err, "error stopping node(s)")
+
+	repository := chain.Nodes()[0].Image.Repository
+	chain.UpgradeVersion(ctx, s.DockerClient, repository, upgradeVersion)
+
+	err = chain.StartAllNodes(ctx)
+	s.Require().NoError(err, "error starting upgraded node(s)")
+
+	// we are reinitializing the clients because we need to update the hostGRPCAddress after
+	// the upgrade and subsequent restarting of nodes
+	s.InitGRPCClients(chain)
+
+	timeoutCtx, timeoutCtxCancel = context.WithTimeout(ctx, time.Minute*2)
+	defer timeoutCtxCancel()
```

Comment:
I'm curious, there's now two defers in this function. Does this reset the context timeout?